File: dsa problems/BinarySearch.java
public class BinarySearch {
    public static int binarySearch(int[] a, int target) {
        int l = 0, r = a.length - 1;
        while (l <= r) {
            int mid = l + (r - l) / 2;
            if (a[mid] == target) return mid;
            else if (a[mid] < target) l = mid + 1;
            else r = mid - 1;
        }
        return -1;
    }

    public static void main(String[] args) {
        int[] a = {1,2,3,4,5,6};
        System.out.println(binarySearch(a, 4)); // expected 3
        System.out.println(binarySearch(a, 7)); // expected -1
    }
}

File: PYTHON FUNDAMENTAL/binary search.py
#Python program to implement binary search 

def binary_search(arr, target):
    low, high = 0, len(arr) - 1
    
    while low <= high:
        mid = (low + high) // 2  # Find the middle index
        mid_element = arr[mid]
        
        if mid_element == target:
            return mid  # Target found, return index
        elif mid_element < target:
            low = mid + 1  # Search in the right half
        else:
            high = mid - 1  # Search in the left half
    
    return -1  # Target not found

# Example usage
array = [int(x) for x in input("Enter sorted numbers separated by spaces: ").split()]
target = int(input("Enter the number to search: "))

result = binary_search(array, target)

if result != -1:
    print(f"{target} found at index {result}")
else:
    print(f"{target} not found in the array")
